bakery/checkers/checkers.go

```diff
@@ -4,192 +4,187 @@ package checkers
+// CheckFirstPartyCaveat implements bakery.FirstPartyCaveatChecker
+// by checking the caveat against all registered caveats conditions.
+func (c *Checker) CheckFirstPartyCaveat(ctxt context.Context, cav string) error {
 	cond, arg, err := ParseCaveat(cav)
 	if err != nil {
 		// If we can't parse it, perhaps it's in some other format,
 		// return a not-recognised error.
 		return errgo.WithCausef(err, ErrCaveatNotRecognized, "cannot parse caveat %q", cav)
 	}
-	if err := c.Check(cond, arg); err != nil {
+	cf, ok := c.checkers[cond]
+	if !ok {
+		return errgo.NoteMask(ErrCaveatNotRecognized, fmt.Sprintf("caveat %q not satisfied", cav), errgo.Any)
+	}
+	if err := cf.Check(ctxt, cond, arg); err != nil {
 		return errgo.NoteMask(err, fmt.Sprintf("caveat %q not satisfied", cav), errgo.Any)
 	}
 	return nil
 }
 
-// TODO add multiChecker.CheckThirdPartyCaveat ?
-// i.e. make this stuff reusable for 3rd party caveats too.
-
-func firstParty(cond, arg string) Caveat {
-	return Caveat{
-		Condition: cond + " " + arg,
-	}
-}
-
-// CheckerFunc implements Checker for a function.
-type CheckerFunc struct {
-	// Condition_ holds the condition that the checker
-	// implements.
-	Condition_ string
-
-	// Check_ holds the function to call to make the check.
-	Check_ func(cond, arg string) error
-}
-
-// Condition implements Checker.Condition.
-func (f CheckerFunc) Condition() string {
-	return f.Condition_
-}
+var errBadCaveat = errgo.New("bad caveat")
 
-// Check implements Checker.Check
-func (f CheckerFunc) Check(cond, arg string) error {
-	return f.Check_(cond, arg)
+func checkError(ctxt context.Context, _, arg string) error {
+	return errBadCaveat
 }
 
-// Map is a checker where the various checkers
-// are specified as entries in a map, one for each
-// condition.
-// The cond argument passed to the function
-// is always the same as its corresponding key
-// in the map.
-type Map map[string]func(cond string, arg string) error
+// ErrCaveatNotRecognized is the cause of errors returned
+// from caveat checkers when the caveat was not
+// recognized.
+var ErrCaveatNotRecognized = errgo.New("caveat not recognized")
 
-// Condition implements Checker.Condition.
-func (m Map) Condition() string {
-	return ""
+// Caveat represents a condition that must be true for a check to
+// complete successfully. If Location is non-empty, the caveat must be
+// discharged by a third party at the given location.
+// The Namespace field holds the namespace URI of the
+// condition - if it is non-empty, it will be converted to
+// a namespace prefix before adding to the macaroon.
+type Caveat struct {
+	Condition string
+	Namespace string
+	Location  string
 }
 
-// Check implements Checker.Check
-func (m Map) Check(cond, arg string) error {
-	f, ok := m[cond]
-	if !ok {
-		return ErrCaveatNotRecognized
-	}
-	if err := f(cond, arg); err != nil {
-		return errgo.Mask(err, errgo.Any)
+func firstParty(cond, arg string) Caveat {
+	return Caveat{
+		Condition: cond + " " + arg,
```

Comment:
Alternatively - you also create a Caveat in httpCaveat with a different namespace, which is currently the only other place cond + " " + args appears. Maybe replace with a NewCaveat(cond, arg string, ns *Namespace)